Pull the owner from the DESCRIBE EXTENDED

Fokko opened the pull request:
I would like to add the owner to the docs, therefore I've written a PR that will do a DESCRIBE EXTENDED to get the relevant fields. Added an unit test, and tested it against Databricks/Azure.

drewbanin commented:
hey @Fokko - can you merge master into this branch? After you do that, I'll kick off the unit + integration tests. I'll give this a deeper look in the coming week :)

Fokko commented:
Sure, I've pulled master into this. Since we have integration testing now as well, I'll check if I can come up with a better test :-)

jtcohen6 commented:
I'm having difficulty testing this branch: `describe extended` does _not_ work in my Spark environment (Databricks on AWS). This results in dbt's `get_catalog` method returning the error, `'NoneType' object has no attribute 'lower'`.

```
describe `my_db`.`my_table`
```
COLUMN_NAME | DATA_TYPE | PK | NULLABLE | DEFAULT | AUTOINCREMENT | COMPUTED | REMARKS | POSITION
-- | -- | -- | -- | -- | -- | -- | -- | --
id | BIGINT | NO | YES |   | NO | NO |   | 1
first_name | STRING(255) | NO | YES |   | NO | NO |   | 2
last_name | STRING(255) | NO | YES |   | NO | NO |   | 3
email | STRING(255) | NO | YES |   | NO | NO |   | 4
gender | STRING(255) | NO | YES |   | NO | NO |   | 5
ip_address | STRING(255) | NO | YES |   | NO | NO |   | 6

```sql
describe extended `my_db`.`my_table`
```
```
No object named extended `my_db`.`my_table` found!
```

It seems like the statement is being parsed wrong, somehow. I wonder if there's some session parameter or cluster configuration I'm missing. I'm running:
```
Databricks Runtime Version 6.2 (includes Apache Spark 2.4.4, Scala 2.11)
```

@Fokko I'm curious if you've encountered a similar issue before?

Fokko commented:
Thanks @jtcohen6 for giving this a try. 

Currently, I'm on holiday so I don't have access to any production environment. But I think you've got an edge case here. With `ANALYZE TABLE my_db.my_table COMPUTE STATISTICS;` you should get the stats. We added this as a post hook:
```
{{ config(
    file_format='delta',
    post_hook=[
        'OPTIMIZE {{ this }}',
        'ANALYZE TABLE {{ this }} COMPUTE STATISTICS'
    ]
) }}
```

If you didn't do this, you'll probably get the error that you're seeing.  Allow me to push a patch.

Fokko commented:
@jtcohen6 Can you share the full stack trace?

jtcohen6 commented:
@Fokko Ah! Sorry to bother you on holiday.

I figured out what's going wrong on my end: `describe extended` _does_ work when I run it from dbt or a Databricks notebook, but not from a SQL runner that connects via [JDBC driver](https://docs.databricks.com/bi/jdbc-odbc-bi.html). The more I know:

<img width="1285" alt="Screen Shot 2020-01-29 at 2 46 12 PM" src="https://user-images.githubusercontent.com/13897643/73392619-2c190e80-42a8-11ea-847d-493f42c85b39.png">

Here's the trace of what's going wrong:
```
2020-01-29 14:54:25,297 (MainThread): 'NoneType' object has no attribute 'lower'
2020-01-29 14:54:25,310 (MainThread): Traceback (most recent call last):
  File "/Users/jerco/dev/product/dbt-spark/env/lib/python3.7/site-packages/dbt_core-0.14.3-py3.7.egg/dbt/main.py", line 82, in main
    results, succeeded = handle_and_check(args)
  File "/Users/jerco/dev/product/dbt-spark/env/lib/python3.7/site-packages/dbt_core-0.14.3-py3.7.egg/dbt/main.py", line 151, in handle_and_check
    task, res = run_from_args(parsed)
  File "/Users/jerco/dev/product/dbt-spark/env/lib/python3.7/site-packages/dbt_core-0.14.3-py3.7.egg/dbt/main.py", line 216, in run_from_args
    results = task.run()
  File "/Users/jerco/dev/product/dbt-spark/env/lib/python3.7/site-packages/dbt_core-0.14.3-py3.7.egg/dbt/task/generate.py", line 209, in run
    results = adapter.get_catalog(manifest)
  File "/Users/jerco/dev/product/dbt-spark/dbt/adapters/spark/impl.py", line 184, in get_catalog
    columns += self._parse_relation(relation, table_columns, rel_type, properties)
  File "/Users/jerco/dev/product/dbt-spark/dbt/adapters/spark/impl.py", line 146, in _parse_relation
    logger.info(column.data_type)
  File "/Users/jerco/dev/product/dbt-spark/env/lib/python3.7/site-packages/dbt_core-0.14.3-py3.7.egg/dbt/adapters/base/relation.py", line 356, in data_type
    if self.is_string():
  File "/Users/jerco/dev/product/dbt-spark/env/lib/python3.7/site-packages/dbt_core-0.14.3-py3.7.egg/dbt/adapters/base/relation.py", line 365, in is_string
    return self.dtype.lower() in ['text', 'character varying', 'character',
AttributeError: 'NoneType' object has no attribute 'lower'
```
It looks like [this](https://github.com/fishtown-analytics/dbt/blob/dev/0.15.2/core/dbt/adapters/base/column.py#L52) is the offending line. I'm not sure if it's because:
* We might be passing `None` as the `dtype` (the empty line between the last column and `# Detailed Table Information`? )
* Spark's string datatype is actually called `string`, which is not in the default implementation. I know that the BigQuery adapter has to reimplement the `Column` class [for similar reasons](https://github.com/fishtown-analytics/dbt/blob/dev/0.15.2/plugins/bigquery/dbt/adapters/bigquery/column.py#L100)

jtcohen6 commented:
@drewbanin Would you be able to take a look at the error in `_parse_relation` related to `None` data type? Once we can get `dbt docs generate` to work, I think this PR is in solid shape.

Fokko commented:
I think the empty line is being transformed into a None, I can do additional checks today or tomorrow.

drewbanin commented:
I think the issue here is indeed that `describe extended` returns a whole lot of information besides just the column names + types. While @Fokko is correct above that the issue is the blank line in the result set, there's also going to be an issue around all of these other metadata fields!

<img width="878" alt="Screen Shot 2020-02-06 at 10 27 28 PM" src="https://user-images.githubusercontent.com/1541139/73998548-fa4f1a00-492f-11ea-8227-b970a402e824.png">

In `spark__get_columns_in_relation`, it's not enough to just add `extended` to the `describe` statement. Instead, I think we should rip out the ensuing code and replace it with something more sensible:

```
{% macro spark__get_columns_in_relation(relation) -%}
  {% call statement('get_columns_in_relation', fetch_result=True) %}
    describe extended {{ relation }}
  {% endcall %}

  {% set table = load_result('get_columns_in_relation').table %}
  -- TODO : Remove this line
  -- {{ return(sql_convert_columns_in_relation(table)) }}

  -- TODO : Call an adapter method (exposed in the compilation context)
  --        which will convert the raw `describe extended` results into a list of columns
  {{ return(adapter.parse_describe_extended(table) }}

{% endmacro %}
```

For reference the dbt Core 0.14.3 implementation of `sql_convert_columns_in_relation` is:

https://github.com/fishtown-analytics/dbt/blob/b5aff36253f5c6563e942265bbaa4cb366722b14/core/dbt/include/global_project/macros/adapters/common.sql#L109-L115

Ultimately, this logic is going to look pretty similar to the current contents of `_parse_relation` in this PR. I think they key difference will be that we should parse the raw results of the `describe extended` statement before wrapping those results up in `Column` objects.

drewbanin commented on dbt/adapters/spark/impl.py:
```diff
@@ -90,53 +107,79 @@ def drop_relation(self, relation, model_name=None):
-    def get_catalog(self, manifest):
-        schemas = manifest.get_used_schemas()
+    @staticmethod
+    def _parse_relation(relation: Relation,
```
There's a lot going on in this method! Can we break this into two (or more) methods that are a little more self-contained?

We should have at least one method, `parse_describe_extended_results`, or similar. This method takes the agate table response from `describe extended...` and returns a data structure containing a list of columns, plus an arbitrary key/value dict of properties returned by the statement. From there, other methods can operate on a proper list of Column objects, plus any properties parsed out from the `describe extended` call as needed.

jtcohen6 commented:
@Fokko Thank you for making these changes! This PR worked for me when I checked it out last night. I'm kicking off tests now.

Fokko commented:
Thanks Drew for the pointers, this made the code a bit nicer.

I'm unable to change the `OWNER` in Databricks. Last December this was possible, maybe something changed. Right now, it is stuck at `root`:
![image](https://user-images.githubusercontent.com/1134248/74664056-45bdbf80-519d-11ea-9883-df09a485b97c.png)

I can check against AWS EMR to see if this is possible there. I've tested against Azure Databricks, and the `DESCRIBE TABLE EXTENDED` is properly working.

Fokko commented:
My pleasure @jtcohen6. Let me know what comes out of it.

Fokko commented:
@jtcohen6 The error looks unrelated. After this one, I'll continue on two other PR's:

- Add support for Table properties. So you can also set the owner of a table through the `Owner` property: https://github.com/fishtown-analytics/dbt-spark/pull/39/files#diff-5cd7d7f886de03e2de4b8cd030ee8c43R151-R156
- Add support for fetching the statistics. I have a first draft in https://github.com/fishtown-analytics/dbt-spark/pull/41 but this needs TLC.

jtcohen6 commented:
@Fokko I figured out why the integration test is failing!

* [Here](https://github.com/fishtown-analytics/dbt-integration-tests/blob/spark-support/features/001_basic_materializations.feature#L49) the test uses the `dbt_utils.equality` macro.
* [Here](https://github.com/fishtown-analytics/dbt-utils/blob/master/macros/schema_tests/equality.sql#L12) is where that macro expects that `adapter.get_columns_in_relation` returns an array of column dicts

Right now, the implementation of `parse_describe_extended` does not include a `quoted` attribute. On Spark, it's basically just `` `column_name` ``.

Fokko commented:
Thanks for figuring that out @jtcohen6. I've introduced the SparkColumn and cleaned up the code a bit. Tested with some simple workflows against Azure Databricks. Also tested the docs and those look good as well.

jtcohen6 commented on test/unit/test_adapter.py:
```diff
@@ -87,3 +87,80 @@ def hive_thrift_connect(host, port, username):
+        input_cols = [Row(keys=['col_name', 'data_type'], values=r) for r in plain_rows]
+
+        config = self.get_target_http(self.project_cfg)
+        rows = SparkAdapter(config).parse_describe_extended(relation, input_cols)
```
This test is failing because you're no longer defining `parse_describe_extended`

Fokko replied:
Oops, I was testing against an actual cluster, so I forgot the unit tests

jtcohen6 commented on dbt/adapters/spark/impl.py:
```diff
@@ -108,53 +115,64 @@ def drop_relation(self, relation, model_name=None):
-
-        return dbt.clients.agate_helper.table_from_data(columns, column_names)
+                logger.debug("Getting table schema for relation {}", relation)
+                columns += list(map(to_dict, self.get_columns_in_relation(relation)))
```
There's only one `flake8` error, and it's on this line:
```
dbt/adapters/spark/impl.py:176:80: E501 line too long (85 > 79 characters)
```

Fokko replied:
Whoops, thanks!

jtcohen6 commented:
@drewbanin I'm unable to locally replicate the latest integration test failure. Could you take a look when you have a second?

Fokko commented:
@jtcohen6 I was already looking into it. Not sure what caused it, could it be a race condition? This could also be wishful thinking.
Also a bit weird that we don't have any logs. Curious about the SQL statements.

Fokko commented:
@jtcohen6 Can we trigger the tests once more? :)

jtcohen6 commented:
@Fokko :( I'll try to get to the bottom of this

drewbanin commented:
@Fokko @jtcohen6 I gave this a spin locally and I found that the two tables (once created via a seed, one created via an incremental model) were created with the columns in a weird order. 

This manifests as an error when the incremental model is run twice. It writes columns out of order, which looks something like this:

![Screen Shot 2020-03-05 at 5 28 16 PM](https://user-images.githubusercontent.com/1541139/76031934-c18e6a80-5f06-11ea-8429-768084d53f45.png)

Have you two seen anything like this before?

Fokko commented:
I got the integration tests working locally as well. I've noticed that the inserted data is shifted by one, the `id` column is prepended to the statement.

It looks like the SEED is not properly being created:
```
create table reporting.seed (id BIGINT,first_name STRING,last_name STRING,email STRING,gender STRING,ip_address STRING)
```
the `id` should be at the end

jtcohen6 commented on dbt/adapters/spark/impl.py:
```diff
@@ -108,53 +115,71 @@ def drop_relation(self, relation, model_name=None):
+    def find_table_information_separator(rows: List[dict]) -> int:
+        pos = 0
+        for row in rows:
+            if not row['col_name']:
```
To avoid including partition columns twice, could this be something like
```suggestion
            if not row['col_name'] or row['col_name'] == '# Partition Information':
```
I'm not sure if that's the most elegant way of writing, and it would need at least a line break to satisfy pep8

jtcohen6 commented:
@Fokko Could you take a look at this when you get a chance? There's one small change I think we should make here to avoid duplicate columns in the `get_columns_in_relation` result.

Fokko commented:
Thanks for pining me @jtcohen6. I was away last week, and the notification got lost. I've updated the PR

jtcohen6 commented on test/unit/test_adapter.py:
```diff
@@ -87,3 +87,86 @@ def hive_thrift_connect(host, port, username):
+
+        config = self.get_target_http(self.project_cfg)
+        rows = SparkAdapter(config).parse_describe_extended(relation, input_cols)
+        self.assertEqual(len(rows), 3)
```
I think the change you pushed will now exclude the third plain row from the `parse_describe_extended` output
```suggestion
        self.assertEqual(len(rows), 2)
```

jtcohen6 commented on test/unit/test_adapter.py:
```diff
@@ -87,3 +87,86 @@ def hive_thrift_connect(host, port, username):
+            'numeric_scale': None,
+            'numeric_precision': None,
+            'char_size': None
+        })
```
I think you'll need to remove this assertion as well

Fokko commented:
Yes, just checked to be sure:
![image](https://user-images.githubusercontent.com/1134248/76791378-0fd22200-67c1-11ea-9109-9b1587e39110.png)

Fokko commented:
We can test against Spark when https://github.com/fishtown-analytics/dbt-spark/pull/60 has been merged :)

jtcohen6 commented:
Thanks @Fokko!

As it is, #39 depends on #60 _and_ #60 depends on #39 for passing integration tests. I've confirmed locally that the two in combination can pass this integration test.

I think I'm going to take what I see as the simplest approach here:

1. Squash + merge this PR into `master`
2. Pull these changes into #60
3. Test #60
4. Merge when tests pass

Does that sound ok by you?

Fokko commented:
Sounds good, I'm also happy to cherry-pick https://github.com/fishtown-analytics/dbt-spark/pull/60/ onto this branch.
